### ISSUE TITLE
Emitted decorator metadata is different when --isolatedModules is set

### ISSUE THREAD

**Issue by jpsfs** [Microsoft/TypeScript#8355]
Hi,

I'm facing an issue while compiling an Angular 2 application with the IsolatedModules flag set to true.
If IsolatedModules is set to false everything works as expect.

The emitted decorator metadata is different. Is this by design or a bug?

My problem is that my module is getting bigger by the day and using this flag I can greatly improve the CompileOnSave experience.

**TypeScript Version:** 

1.8.7

**Code**

```ts
import * as ng from "angular2/core";

@ng.Decorator({
...
})
export class MyClass{
     constructor(private el: ng.ElementRef) {
     }
}
```

**Expected behavior:**
If the isolatedModules flag is set to false:
```js
...
MyClass = __decorate([
      ng.Directive({
            selector: '[my-selector]',
      }), 
      __metadata('design:paramtypes', [ng.ElementRef]) // <<<<<<<<<<<<<<<<<<<
], MyClass);
```

**Actual behavior:** 

If the isolatedModules flag is set to true:

```js
...
MyClass = __decorate([
      ng.Directive({
            selector: '[my-selector]',
      }), 
      __metadata('design:paramtypes', [Object]) // <<<<<<<<<<<<<<<<<<<<<<<
], MyClass);
```

**Comment by mhegazy** [comment]
there are differences that are expected in the emit when using this flag, but it should not be functionally different. this sample looks like a bug. 

I can not get the different emitted code just using the snippet provided, so there is something else going on. can you share more context from this file.

**Comment by jpsfs** [comment]
Sorry, it seems that it is a problem within gulp-typescript and not typescript itself.
I'm closing the issue for now.

Nevertheless, here is a repo that can reproduce the bug:
https://github.com/jpsfs/typescript-angular-samples

**Comment by ivogabe** [comment]
@mhegazy I'm confused, has the behavior of `isolatedModules` been changed? I thought that the project would be compiled file by file, like what `transpile` does. But that doesn't seem to be what `tsc` really does?

**Comment by mhegazy** [comment]
The output should be the same as that of transpile. if not it is an error.

**Comment by ivogabe** [comment]
Gulp-typescript uses transpile, and that emits `Object`, whereas tsc emits `ng.ElementRef`. So I think this would be a bug of TypeScript, as I have no idea why gulp-typescript would give a different result otherwise.

**Comment by yuit** [comment]
@andy-ms you mind if I take this bug

**Comment by andy-ms** [comment]
OK. Here's what I was testing with:

    // @experimentalDecorators: true
    // @emitDecoratorMetadata: true
    // @target: es5
    // @isolatedModules: true

    // @Filename: a.ts
    export interface X {}

    // @Filename: user.ts
    import {X} from './a'
    declare const decorator: MethodDecorator
    export class C {
        @decorator
        myMethod(x: X) {}
    }

When `a.ts` declares `X` as an interface, the parameter type is output as `Object`, but when `X` is a class, it is emitted as `a_1.X`.
So it appears that even with `--isolatedModules`, we are looking at information from a different module to affect emit.

**Comment by rbuckton** [comment]
@andy-ms That is expected. We always serialize an interface as Object.

**Comment by yuit** [comment]
@rbuckton is this because interface has no value that one can refer to during run-time ?

**Comment by yuit** [comment]
@ivogabe so there is some inconsistency in our emitted results for sure. I am currently investigating. Though some quick clarification, just wonder why `--isolatedModules` specified by users make gulp-typescript behaves different since you guys call into TypeScript' api transpile function [here](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L235) I think I may misunderstand something here...

**Comment by ivogabe** [comment]
Gulp-typescript has two different compile strategies. When you set `isolatedModules`, it will use a file based compiler, by calling the transpile function. With incremental compilation this is very fast, since only changed files will be recompiled. If you don't set `isolatedModules`, it will use a project based compiler. This works with a CompilerHost etcetera, and all files are passed to the compiler.
`tsc` always takes that second approach.

**Comment by ivogabe** [comment]
The problem is probably that only type information of the containing file is present with `transpile`. `tsc` does have type information of all source files but it shouldn't use that in the emit.
